include/spblas/vendor/aoclsparse/spgemm_impl.hpp

```
@@ -0,0 +1,199 @@
+  log_trace("");
+  auto a_base = __detail::get_ultimate_base(a);
+  auto b_base = __detail::get_ultimate_base(b);
+
+  using T = tensor_scalar_t<C>;
+  using I = tensor_index_t<C>;
+  using O = tensor_offset_t<C>;
+
+  aoclsparse_matrix csrA = nullptr;
+  aoclsparse_mat_descr descrA;
+  aoclsparse_status status = aoclsparse_create_mat_descr(&descrA);
+  if (status != aoclsparse_status_success) {
+    fmt::print("\t descr creation failed\n");
+  }
+  aoclsparse_set_mat_type(descrA, aoclsparse_matrix_type_general);
+  aoclsparse_index_base indexingA = aoclsparse_index_base_zero;
+  aoclsparse_operation opA = aoclsparse_operation_none;
+
+  const index_t a_nrows = __backend::shape(a_base)[0];
+  const index_t a_ncols = __backend::shape(a_base)[1];
+
+  aoclsparse_matrix csrB = nullptr;
+  aoclsparse_mat_descr descrB;
+  status = aoclsparse_create_mat_descr(&descrB);
+  if (status != aoclsparse_status_success) {
+    fmt::print("\t descr creation failed\n");
+  }
+
+  aoclsparse_set_mat_type(descrB, aoclsparse_matrix_type_general);
+  aoclsparse_index_base indexingB = aoclsparse_index_base_zero;
+  aoclsparse_operation opB = aoclsparse_operation_none;
+
+  const index_t b_nrows = __backend::shape(b_base)[0];
+  const index_t b_ncols = __backend::shape(b_base)[1];
+
+  aoclsparse_matrix csrC = nullptr;
+  aoclsparse_index_base indexingC = aoclsparse_index_base_zero;
+  index_t c_nrows = 0, c_ncols = 0;
+  offset_t c_nnz = -1;
+  offset_t* c_rowptr = nullptr;
+  index_t* c_colind = nullptr;
+  T* c_values = nullptr;
+
+  const aoclsparse_int nnzA = a_base.rowptr().data()[a_nrows];
+  const aoclsparse_int nnzB = b_base.rowptr().data()[b_nrows];
+
+  status = __aoclsparse::aoclsparse_create_csr(
+      &csrA, indexingA, a_nrows, a_ncols, nnzA, a_base.rowptr().data(),
+      a_base.colind().data(), a_base.values().data());
+  if (status != aoclsparse_status_success) {
+    fmt::print("\t csr matrix A creation failed\n");
+  }
```

do we have a need to exit if there is a failure here instead of keeping going ?

---

I would say the thing to do here is throw an exception.  (And also to probably clean up any temporary objects that have been created.)

---

The oneMKL SYCL backend will throw exceptions on failure, and that's the only failure mode, right?  I think it's probably fine to let those exceptions go through to the user, although I guess we could also try to clean up and re-throw new `spblas` exceptions.  (We haven't talked about specific exceptions to throw, but the standard thing to do is to come up with some inheriting from/based on the [standard exception categories](https://en.cppreference.com/w/cpp/error#Exception_categories).)

---

We should handle failures instead of ignoring it. Also, it would be nice to have consistent handling across vendor backends.  Re-throwing new spblas exceptions is a good idea, something to discuss in future meetings. As most of these functions return "void", returning a status is not possible as is, right?